File: app/src/main/java/com/futuretech/nfmovies/Adapter/ChannelGridViewAdapter.java
package com.futuretech.nfmovies.Adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.cardview.widget.CardView;
import com.futuretech.nfmovies.Entity.Channel;
import com.futuretech.nfmovies.Listener.ChannelClickListener;
import com.futuretech.nfmovies.R;
import com.futuretech.nfmovies.Utils.ImageUtil;

import java.util.List;

public class ChannelGridViewAdapter extends BaseAdapter {
    private List<Channel> channels;
    private Context context;
    private ChannelClickListener clickListener;

    public ChannelGridViewAdapter(Context context, List<Channel> channels, ChannelClickListener channelClickListener) {
        this.channels = channels;
        this.context = context;
        this.clickListener = channelClickListener;
    }

    public ChannelGridViewAdapter(Context context, ChannelClickListener channelClickListener) {
        this.context = context;
        this.channels = null;
        this.clickListener = channelClickListener;
    }

    public List<Channel> getChannels() {
        return channels;
    }

    @Override
    public int getCount() {
        return channels.size();
    }

    @Override
    public Object getItem(int position) {
        return channels.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder;
        if (convertView == null) {
            convertView = LayoutInflater.from(context).inflate(R.layout.channel_item, null);
            viewHolder = new ViewHolder(convertView);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }
        viewHolder.channelTitle.setText(channels.get(position).getTitle());
        ImageUtil.display(context, channels.get(position).getIcon(), viewHolder.channelImg, null);
        viewHolder.channelImg.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clickListener.OnChannelClick(channels.get(position));
            }
        });
        convertView.setTag(viewHolder);
        return convertView;
    }

    class ViewHolder {
        private CardView cardView;
        private ImageView channelImg;
        private TextView channelTitle;
        public ViewHolder(View convertView) {
            channelImg = convertView.findViewById(R.id.channel_img);
            channelTitle = convertView.findViewById(R.id.channel_title);
            cardView = convertView.findViewById(R.id.channel_card);
        }
    }
}

File: app/src/main/java/com/futuretech/nfmovies/Adapter/ChannelGridViewAdapter.kt
package com.futuretech.nfmovies.Adapter

import android.content.Context
import android.graphics.Bitmap
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Transformation
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import com.futuretech.nfmovies.Entity.Channel
import com.futuretech.nfmovies.Listener.ChannelClickListener
import com.futuretech.nfmovies.R
import com.futuretech.nfmovies.Utils.ImageUtil

class ChannelGridViewAdapter : BaseAdapter {
    var channels: List<Channel>? = null
        private set
    private var context: Context? = null
    private var clickListener: ChannelClickListener? = null

    constructor(context: Context, channels: List<Channel>, channelClickListener: ChannelClickListener) {
        this.channels = channels
        this.context = context
        this.clickListener = channelClickListener
    }

    constructor(context: Context, channelClickListener: ChannelClickListener) {
        this.context = context
        this.channels = null
        this.clickListener = channelClickListener
    }

    override fun getCount(): Int {
        return channels!!.size
    }

    override fun getItem(position: Int): Any {
        return channels!![position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertview: View?, parent: ViewGroup): View {
        var convertView = convertview
        val viewHolder: ViewHolder
        if (convertView == null) {
            convertView = LayoutInflater.from(context).inflate(R.layout.channel_item, null)
            viewHolder = ViewHolder(convertView!!)
        } else {
            viewHolder = convertView.tag as ViewHolder
        }
        viewHolder.channelTitle.text = channels!![position].title
        ImageUtil.display(context!!, channels!![position].icon, viewHolder.channelImg, null)
        viewHolder.channelImg.setOnClickListener { clickListener!!.OnChannelClick(channels!![position]) }
        convertView.tag = viewHolder
        return convertView
    }

    internal inner class ViewHolder(convertView: View) {
        private val cardView: CardView
        val channelImg: ImageView
        val channelTitle: TextView

        init {
            channelImg = convertView.findViewById(R.id.channel_img)
            channelTitle = convertView.findViewById(R.id.channel_title)
            cardView = convertView.findViewById(R.id.channel_card)
        }
    }
}
